## 3.4. ANN

In [1]:
#https://machinelearningmastery.com/binary-classification-tutorial-with-the-keras-deep-learning-library/
# https://machinelearningmastery.com/grid-search-hyperparameters-deep-learning-models-python-keras/
import numpy
import pandas as pd
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import StratifiedKFold
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score, cohen_kappa_score
from sklearn.metrics import roc_curve, auc
from sklearn.metrics import log_loss
import matplotlib.pyplot as plt
from sklearn.metrics import roc_auc_score

# Use scikit-learn to grid search the batch size and epochs
from sklearn.model_selection import GridSearchCV
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier

Using TensorFlow backend.
/Users/sdoneva/anaconda/lib/python3.6/importlib/_bootstrap.py:205: RuntimeWarning: compiletime version 3.5 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.6
  return f(*args, **kwds)


In [2]:
df=pd.read_csv('final_train_data_StdSc.csv')

In [3]:
df_test=pd.read_csv('final_test_data_StdSc.csv')

In [4]:
# Get all the columns from the dataframe.
columns = df.columns.tolist()
# Filter the columns to remove ones we don't want.
columns_subset = [c for c in columns if c not in ["is_churn","msno","Unnamed: 0","Unnamed: 0.1"]]
# Store the variable we'll be predicting on.
target = "is_churn"

In [5]:
X_train = df[columns_subset]
y_train = df[target]

X_train.shape

(70598, 154)

In [6]:
y_train=df[target]
y_train.shape

(70598,)

In [7]:
X_test = df_test[columns_subset]
y_test=df_test[target]

X_test.shape

(262843, 154)

### Baseline model

In [8]:
def baseline_model(features):
    # create model
    model = Sequential()
    model.add(Dense(100, input_dim=features, kernel_initializer='normal', activation='relu'))
    #if we want a hidden layer :
    #model.add(Dense(8, activation='relu'))
    model.add(Dense(1, kernel_initializer='normal', activation='sigmoid'))
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

### Decide on dimensionality

### All features

In [52]:
model_all=baseline_model(154)
model_all.fit(X_train, y_train,epochs=10, batch_size=16, verbose=1)

Epoch 1/10
70598/70598 [==============================] - 9s 127us/step - loss: 0.2474 - acc: 0.9087
Epoch 2/10
70598/70598 [==============================] - 9s 123us/step - loss: 0.2182 - acc: 0.9211
Epoch 3/10
70598/70598 [==============================] - 8s 120us/step - loss: 0.2118 - acc: 0.9227
Epoch 4/10
70598/70598 [==============================] - 9s 125us/step - loss: 0.2082 - acc: 0.9250
Epoch 5/10
70598/70598 [==============================] - 9s 120us/step - loss: 0.2052 - acc: 0.9250
Epoch 6/10
70598/70598 [==============================] - 9s 127us/step - loss: 0.2021 - acc: 0.9255
Epoch 7/10
70598/70598 [==============================] - 9s 127us/step - loss: 0.1997 - acc: 0.9264
Epoch 8/10
70598/70598 [==============================] - 9s 126us/step - loss: 0.1971 - acc: 0.9268
Epoch 9/10
70598/70598 [==============================] - 9s 130us/step - loss: 0.1951 - acc: 0.9271
Epoch 10/10
70598/70598 [==============================] - 9s 124us/step - loss: 0.1930 - a

In [53]:
analytics(y_test,model_all,X_test)

Confusion Matrix:
      Prediction
         0     1
    0220938 24612
    1 1425 15868

Accuracy Score:
0.90094086584
Classification Report:
             precision    recall  f1-score   support

          0       0.99      0.90      0.94    245550
          1       0.39      0.92      0.55     17293

avg / total       0.95      0.90      0.92    262843

Log Loss:
0.225461077679
AUC Score:
0.96204957219


### Top 59

In [54]:
top_59=['number_of_days_201702_listened','is_auto_renew','total_cancel','active_days','avg_plan_list_price','avg_actual_amount_paid','num_unq_201702_sum','num_totalsec_201702_sum','number_of_days_lasttwo_listened','most_fq_payment_method_id','num_75_mean','number_of_days_listened','num_25_201702_sum','num_50_201702_sum','num_75_201702_sum','num_100_201702_sum','num_25_201702_mean','num_50_201702_mean','num_100_lasttwo_mean','num_totalsec_lasttwo_mean','number_of_days_lastthree_listened','total_churn','registered_via_3','registered_via_4','registered_via_7','bd','num_100_sum','num_totalsec_sum','num_50_mean','num_100_mean','num_totalsec_mean','num_100_med','num_totalsec_med','num_25_max','num_unq_max','num_totalsec_max','num_totalsec_min','num_985_201702_sum','num_75_201702_mean','num_100_201702_mean','num_unq_201702_mean','num_totalsec_201702_mean','num_unq_201701_sum','num_100_201701_mean','number_of_days_201701_listened','num_25_201612_mean','num_100_201612_mean','num_unq_lasttwo_sum','num_75_lasttwo_mean','num_100_lastthree_sum','num_totalsec_lastthree_sum','num_50_lastthree_mean','num_75_lastthree_mean','num_100_lastthree_mean','num_totalsec_lastthree_mean','num_above_50_sum','proportion_songs_above_50','gender_2','registered_via_9']

In [55]:
X_train_59 = df[top_59]
X_test_59 = df_test[top_59]

In [56]:
model_59=baseline_model(59)
model_59.fit(X_train_59, y_train,epochs=10, batch_size=16, verbose=1)

Epoch 1/10
70598/70598 [==============================] - 8s 119us/step - loss: 0.2367 - acc: 0.9133
Epoch 2/10
70598/70598 [==============================] - 8s 115us/step - loss: 0.2120 - acc: 0.9234
Epoch 3/10
70598/70598 [==============================] - 8s 117us/step - loss: 0.2075 - acc: 0.9254
Epoch 4/10
70598/70598 [==============================] - 8s 118us/step - loss: 0.2047 - acc: 0.9264
Epoch 5/10
70598/70598 [==============================] - 9s 124us/step - loss: 0.2017 - acc: 0.9265
Epoch 6/10
70598/70598 [==============================] - 9s 123us/step - loss: 0.1988 - acc: 0.9270
Epoch 7/10
70598/70598 [==============================] - 7s 99us/step - loss: 0.1970 - acc: 0.9277
Epoch 8/10
70598/70598 [==============================] - 8s 115us/step - loss: 0.1948 - acc: 0.9282
Epoch 9/10
70598/70598 [==============================] - 9s 131us/step - loss: 0.1934 - acc: 0.9285
Epoch 10/10
70598/70598 [==============================] - 8s 115us/step - loss: 0.1917 - ac

In [58]:
analytics(y_test,model_59,X_test_59)

Confusion Matrix:
      Prediction
         0     1
    0219013 26537
    1 1052 16241

Accuracy Score:
0.895036200317
Classification Report:
             precision    recall  f1-score   support

          0       1.00      0.89      0.94    245550
          1       0.38      0.94      0.54     17293

avg / total       0.95      0.90      0.91    262843

Log Loss:
0.251584923852
AUC Score:
0.968598060806


### Chosen by RFE

In [59]:
top_RFE=['number_of_days_listened','num_totalsec_201702_sum','num_totalsec_201702_mean','number_of_days_201702_listened','num_unq_201701_sum','number_of_days_201701_listened','num_unq_lasttwo_sum','num_100_lasttwo_mean','num_totalsec_lasttwo_mean','number_of_days_lasttwo_listened','number_of_days_lastthree_listened','most_fq_payment_method_id','is_auto_renew','total_cancel','active_days','avg_plan_list_price','avg_actual_amount_paid','registered_via_3','registered_via_4','registered_via_9']

In [60]:
X_train_RFE = df[top_RFE]
X_test_RFE = df_test[top_RFE]

In [61]:
model_rfe=baseline_model(20)
model_rfe.fit(X_train_RFE, y_train,epochs=10, batch_size=16, verbose=1)

Epoch 1/10
70598/70598 [==============================] - 8s 118us/step - loss: 0.2317 - acc: 0.9163
Epoch 2/10
70598/70598 [==============================] - 8s 108us/step - loss: 0.2104 - acc: 0.9244
Epoch 3/10
70598/70598 [==============================] - 8s 115us/step - loss: 0.2066 - acc: 0.9252
Epoch 4/10
70598/70598 [==============================] - 10s 136us/step - loss: 0.2036 - acc: 0.9260
Epoch 5/10
70598/70598 [==============================] - 9s 126us/step - loss: 0.2007 - acc: 0.9273
Epoch 6/10
70598/70598 [==============================] - 8s 109us/step - loss: 0.1983 - acc: 0.9275
Epoch 7/10
70598/70598 [==============================] - 8s 108us/step - loss: 0.1966 - acc: 0.9282
Epoch 8/10
70598/70598 [==============================] - 8s 106us/step - loss: 0.1950 - acc: 0.9280
Epoch 9/10
70598/70598 [==============================] - 8s 107us/step - loss: 0.1940 - acc: 0.9290
Epoch 10/10
70598/70598 [==============================] - 8s 115us/step - loss: 0.1931 - 

In [62]:
analytics(y_test,model_rfe,X_test_RFE)

Confusion Matrix:
      Prediction
         0     1
    0220387 25163
    1  927 16366

Accuracy Score:
0.900739224556
Classification Report:
             precision    recall  f1-score   support

          0       1.00      0.90      0.94    245550
          1       0.39      0.95      0.56     17293

avg / total       0.96      0.90      0.92    262843

Log Loss:
0.214808279573
AUC Score:
0.970122542678


### Chosen by PCA

In [63]:
top_PCA=['num_100_mean','num_100_lastthree_mean','num_totalsec_mean','num_totalsec_lastthree_mean','num_100_med','num_100_lasttwo_mean','num_totalsec_med','num_totalsec_lasttwo_mean','num_75_lastthree_mean','num_50_lastthree_mean','num_100_lastthree_sum','num_100_sum','num_totalsec_lastthree_sum','num_75_mean','num_totalsec_sum','num_above_50_sum','num_100_201701_mean','num_100_201612_mean','num_50_mean','num_75_lasttwo_mean']

In [64]:
X_train_PCA = df[top_PCA]
X_test_PCA = df_test[top_PCA]

In [65]:
model_pca=baseline_model(20)
model_pca.fit(X_train_PCA, y_train,epochs=10, batch_size=16, verbose=1)

Epoch 1/10
70598/70598 [==============================] - 10s 139us/step - loss: 0.6715 - acc: 0.5787
Epoch 2/10
70598/70598 [==============================] - 8s 118us/step - loss: 0.6645 - acc: 0.5927
Epoch 3/10
70598/70598 [==============================] - 8s 114us/step - loss: 0.6618 - acc: 0.5969
Epoch 4/10
70598/70598 [==============================] - 7s 96us/step - loss: 0.6604 - acc: 0.6006
Epoch 5/10
70598/70598 [==============================] - 6s 89us/step - loss: 0.6594 - acc: 0.6012
Epoch 6/10
70598/70598 [==============================] - 7s 94us/step - loss: 0.6581 - acc: 0.6033
Epoch 7/10
70598/70598 [==============================] - 7s 105us/step - loss: 0.6571 - acc: 0.6042
Epoch 8/10
70598/70598 [==============================] - 8s 118us/step - loss: 0.6559 - acc: 0.6080
Epoch 9/10
70598/70598 [==============================] - 9s 126us/step - loss: 0.6545 - acc: 0.6097
Epoch 10/10
70598/70598 [==============================] - 8s 114us/step - loss: 0.6538 - acc

In [66]:
analytics(y_test,model_pca,X_test_PCA)

Confusion Matrix:
      Prediction
         0     1
    0176298 69252
    1 8988  8305

Accuracy Score:
0.702331810244
Classification Report:
             precision    recall  f1-score   support

          0       0.95      0.72      0.82    245550
          1       0.11      0.48      0.18     17293

avg / total       0.90      0.70      0.78    262843

Log Loss:
0.645068196714
AUC Score:
0.646011762274


### Chosen by F-classification

In [67]:
top_F=['num_25_201702_sum','num_50_201702_sum','num_75_201702_sum','num_985_201702_sum','num_100_201702_sum','num_unq_201702_sum','num_totalsec_201702_sum','number_of_days_201702_listened','number_of_days_lasttwo_listened','number_of_days_lastthree_listened','total_churn','is_auto_renew','total_cancel','active_days','avg_plan_list_price','avg_actual_amount_paid','gender_2','registered_via_3','registered_via_4','registered_via_7']

In [68]:
X_train_F = df[top_F]
X_test_F = df_test[top_F]

In [69]:
model_f=baseline_model(20)
model_f.fit(X_train_F, y_train,epochs=10, batch_size=16, verbose=1)

Epoch 1/10
70598/70598 [==============================] - 8s 114us/step - loss: 0.2362 - acc: 0.9151
Epoch 2/10
70598/70598 [==============================] - 8s 118us/step - loss: 0.2132 - acc: 0.9243
Epoch 3/10
70598/70598 [==============================] - 7s 104us/step - loss: 0.2093 - acc: 0.9256
Epoch 4/10
70598/70598 [==============================] - 7s 104us/step - loss: 0.2054 - acc: 0.9262
Epoch 5/10
70598/70598 [==============================] - 7s 105us/step - loss: 0.2027 - acc: 0.9279
Epoch 6/10
70598/70598 [==============================] - 8s 107us/step - loss: 0.2004 - acc: 0.9281
Epoch 7/10
70598/70598 [==============================] - 7s 95us/step - loss: 0.1986 - acc: 0.9287
Epoch 8/10
70598/70598 [==============================] - 7s 105us/step - loss: 0.1970 - acc: 0.9291
Epoch 9/10
70598/70598 [==============================] - 7s 106us/step - loss: 0.1957 - acc: 0.9296
Epoch 10/10
70598/70598 [==============================] - 7s 103us/step - loss: 0.1951 - ac

In [70]:
analytics(y_test,model_f,X_test_F)

Confusion Matrix:
      Prediction
         0     1
    0219464 26086
    1  811 16482

Accuracy Score:
0.897668950666
Classification Report:
             precision    recall  f1-score   support

          0       1.00      0.89      0.94    245550
          1       0.39      0.95      0.55     17293

avg / total       0.96      0.90      0.92    262843

Log Loss:
0.210309633582
AUC Score:
0.969964711246


### Chosen by Mutual Info

In [71]:
top_MI=['num_50_201702_sum','num_75_201702_sum','num_100_201702_sum','num_unq_201702_sum','num_totalsec_201702_sum','num_25_201702_mean','num_50_201702_mean','num_75_201702_mean','num_100_201702_mean','num_unq_201702_mean','number_of_days_201702_listened','number_of_days_lasttwo_listened','most_fq_payment_method_id','total_churn','is_auto_renew','total_cancel','active_days','avg_plan_list_price','avg_actual_amount_paid','registered_via_7']

In [72]:
X_train_MI = df[top_MI]
X_test_MI = df_test[top_MI]

In [73]:
model_mi=baseline_model(20)
model_mi.fit(X_train_MI, y_train,epochs=10, batch_size=16, verbose=1)

Epoch 1/10
70598/70598 [==============================] - 8s 116us/step - loss: 0.2321 - acc: 0.9158
Epoch 2/10
70598/70598 [==============================] - 7s 106us/step - loss: 0.2101 - acc: 0.9251
Epoch 3/10
70598/70598 [==============================] - 7s 95us/step - loss: 0.2063 - acc: 0.9256
Epoch 4/10
70598/70598 [==============================] - 8s 107us/step - loss: 0.2038 - acc: 0.9267
Epoch 5/10
70598/70598 [==============================] - 9s 130us/step - loss: 0.2012 - acc: 0.9267
Epoch 6/10
70598/70598 [==============================] - 10s 147us/step - loss: 0.1993 - acc: 0.9273
Epoch 7/10
70598/70598 [==============================] - 7s 102us/step - loss: 0.1970 - acc: 0.9278
Epoch 8/10
70598/70598 [==============================] - 7s 98us/step - loss: 0.1957 - acc: 0.9282
Epoch 9/10
70598/70598 [==============================] - 7s 95us/step - loss: 0.1943 - acc: 0.9281
Epoch 10/10
70598/70598 [==============================] - 7s 95us/step - loss: 0.1935 - acc:

In [74]:
analytics(y_test,model_mi,X_test_MI)

Confusion Matrix:
      Prediction
         0     1
    0219656 25894
    1  880 16413

Accuracy Score:
0.898136910627
Classification Report:
             precision    recall  f1-score   support

          0       1.00      0.89      0.94    245550
          1       0.39      0.95      0.55     17293

avg / total       0.96      0.90      0.92    262843

Log Loss:
0.22791795016
AUC Score:
0.970694025781


### Chosen by random forest

In [75]:
top_RForest=['active_days','is_auto_renew','avg_actual_amount_paid','total_cancel','number_of_days_201702_listened','avg_plan_list_price','most_fq_payment_method_id','num_totalsec_min','number_of_days_listened','bd','num_25_max','num_25_201702_sum','num_unq_max','num_25_201612_mean','num_totalsec_max','proportion_songs_above_50','num_75_mean','num_unq_201702_sum','num_25_201702_mean','num_50_201702_mean']

In [76]:
X_train_Rf = df[top_RForest]
X_test_Rf = df_test[top_RForest]

In [77]:
model_forest=baseline_model(20)
model_forest.fit(X_train_Rf, y_train,epochs=10, batch_size=16, verbose=1)

Epoch 1/10
70598/70598 [==============================] - 8s 114us/step - loss: 0.2332 - acc: 0.9155
Epoch 2/10
70598/70598 [==============================] - 8s 110us/step - loss: 0.2129 - acc: 0.9234
Epoch 3/10
70598/70598 [==============================] - 8s 111us/step - loss: 0.2077 - acc: 0.9258
Epoch 4/10
70598/70598 [==============================] - 8s 114us/step - loss: 0.2040 - acc: 0.9267
Epoch 5/10
70598/70598 [==============================] - 8s 109us/step - loss: 0.2008 - acc: 0.9280
Epoch 6/10
70598/70598 [==============================] - 7s 95us/step - loss: 0.1991 - acc: 0.9277
Epoch 7/10
70598/70598 [==============================] - 7s 95us/step - loss: 0.1974 - acc: 0.9286
Epoch 8/10
70598/70598 [==============================] - 7s 95us/step - loss: 0.1961 - acc: 0.9291
Epoch 9/10
70598/70598 [==============================] - 7s 95us/step - loss: 0.1951 - acc: 0.9293
Epoch 10/10
70598/70598 [==============================] - 7s 95us/step - loss: 0.1943 - acc: 0

In [78]:
analytics(y_test,model_forest,X_test_Rf)

Confusion Matrix:
      Prediction
         0     1
    0220573 24977
    1  967 16326

Accuracy Score:
0.901294689225
Classification Report:
             precision    recall  f1-score   support

          0       1.00      0.90      0.94    245550
          1       0.40      0.94      0.56     17293

avg / total       0.96      0.90      0.92    262843

Log Loss:
0.211072022553
AUC Score:
0.970223929624


### Features without correlation

In [79]:
top_no_corr=['number_of_days_201702_listened','is_auto_renew','total_cancel','active_days','avg_actual_amount_paid','most_fq_payment_method_id','num_75_mean','number_of_days_listened','num_25_201702_sum','num_50_201702_sum','num_100_201702_sum','num_25_201702_mean','num_50_201702_mean','num_totalsec_lasttwo_mean','total_churn','registered_via_3','registered_via_4','registered_via_7','bd','num_totalsec_sum','num_100_med','num_25_max','num_unq_max','num_totalsec_max','num_totalsec_min','num_985_201702_sum','num_75_201702_mean','num_unq_201702_mean','num_unq_201701_sum','num_25_201612_mean','num_100_201612_mean','num_75_lasttwo_mean','num_50_lastthree_mean','proportion_songs_above_50','registered_via_9']

In [80]:
X_train_no_corr = df[top_no_corr]
X_test_no_corr = df_test[top_no_corr]

In [81]:
model_decorr=baseline_model(35)
model_decorr.fit(X_train_no_corr, y_train,epochs=10, batch_size=16, verbose=1)

Epoch 1/10
70598/70598 [==============================] - 7s 106us/step - loss: 0.2339 - acc: 0.9151
Epoch 2/10
70598/70598 [==============================] - 7s 97us/step - loss: 0.2098 - acc: 0.9241
Epoch 3/10
70598/70598 [==============================] - 7s 97us/step - loss: 0.2053 - acc: 0.9258
Epoch 4/10
70598/70598 [==============================] - 7s 99us/step - loss: 0.2028 - acc: 0.9265
Epoch 5/10
70598/70598 [==============================] - 9s 127us/step - loss: 0.2002 - acc: 0.9273
Epoch 6/10
70598/70598 [==============================] - 9s 125us/step - loss: 0.1989 - acc: 0.9276
Epoch 7/10
70598/70598 [==============================] - 9s 125us/step - loss: 0.1976 - acc: 0.9277
Epoch 8/10
70598/70598 [==============================] - 8s 117us/step - loss: 0.1952 - acc: 0.9284
Epoch 9/10
70598/70598 [==============================] - 8s 106us/step - loss: 0.1934 - acc: 0.9286
Epoch 10/10
70598/70598 [==============================] - 8s 107us/step - loss: 0.1918 - acc:

In [82]:
analytics(y_test,model_decorr,X_test_no_corr)

Confusion Matrix:
      Prediction
         0     1
    0222521 23029
    1 1222 16071

Accuracy Score:
0.907735796654
Classification Report:
             precision    recall  f1-score   support

          0       0.99      0.91      0.95    245550
          1       0.41      0.93      0.57     17293

avg / total       0.96      0.91      0.92    262843

Log Loss:
0.205458826098
AUC Score:
0.969268311514


# Summary Statistics

- All features
Log Loss:
0.225461077679
AUC Score:
0.96204957219
F1-1/F1-avg: 0.55/0.92

- Top 59
Log Loss:
0.251584923852
AUC Score:
0.968598060806
F1-1/F1-avg: 0.54/0.92

- RFE	
Log Loss:
0.214808279573
AUC Score:
0.970122542678
F1-1/F1-avg: 0.56/0.92
- PCA
Log Loss:
0.214808279573
AUC Score:
0.970122542678
F1-1/F1-avg: 0.56/0.92
- F-classification
Log Loss:
0.210309633582
AUC Score:
0.969964711246
F1-1/F1-avg: 0.55/0.92
- Mutual Info	
Log Loss:
0.22791795016
AUC Score:
0.970694025781
F1-1/F1-avg: 0.55/0.92
- Random Forrest	
Log Loss:
0.211072022553
AUC Score:
0.970223929624
F1-1/F1-avg: 0.56/0.92
- Without correlation	
Log Loss:
0.205458826098
AUC Score:
0.969268311514
F1-1/F1-avg: 0.57/0.92

>Features after the correlation removal performed best

In [83]:
values_no_corr=model_decorr.predict_proba(X_test_no_corr)
values_df=pd.DataFrame(values_no_corr)
values_df.to_csv('predict_proba_ANN_decorr.csv')

### Sampling 0.1 from train for parameter tuning

In [84]:
X_sample= df.sample(frac=0.1, replace=True)

In [85]:
y_sample=X_sample['is_churn']

In [86]:
X_sample.shape

(7060, 157)

In [87]:
y_sample.sum()

3484

In [88]:
X_sample.head()

Unnamed: 0        bd  num_25_sum  num_50_sum  num_75_sum  num_985_sum  \
16349       16349 -0.811134   -0.455947   -0.685689   -0.502068    -0.445134   
49358       49358  0.754708   -0.708717   -0.785512   -0.841180    -0.741389   
50335       50335  0.629440   -0.818794   -0.839962   -0.900157    -0.705838   
36982       36982 -0.811134   -0.831025   -0.903486   -0.944389    -0.836190   
39963       39963  1.381044   -0.486524   -0.422517   -0.502068    -0.243680   

       num_100_sum  num_unq_sum  num_totalsec_sum  num_25_mean  \
16349     0.349455     0.485696          0.242797    -0.583489   
49358    -0.806527    -0.857370         -0.832914    -0.303228   
50335    -0.852786    -0.937954         -0.881860    -0.719427   
36982    -0.902490    -0.992213         -0.935144    -0.719427   
39963    -0.170712    -0.221831         -0.210368    -0.488205   

                           ...                       gender_0  gender_1  \
16349                      ...                      -0.508672 -0.541352   
49358                      ...                      -0.508672  1.847228   
50335                      ...                       1.965905 -0.541352   
36982                      ...                      -0.508672 -0.541352   
39963                      ...                      -0.508672  1.847228   

       gender_2  registered_via_3  registered_via_4  registered_via_7  \
16349  0.872460         -0.357804         -0.232910          0.863924   
49358 -1.146185         -0.357804          4.293498         -1.157510   
50335 -1.146185          2.794825         -0.232910         -1.157510   
36982  0.872460         -0.357804         -0.232910          0.863924   
39963 -1.146185         -0.357804         -0.232910         -1.157510   

       registered_via_9  registered_via_13  is_churn  \
16349         -0.591173           -0.05888         0   
49358         -0.591173           -0.05888         1   
50335         -0.591173           -0.05888         1   
36982         -0.591173           -0.05888         1   
39963          1.691552           -0.05888         1   

                                               msno  
16349  5R8fn3LSz6IAAWr8i+10VpQUGWRMvslhSMQgG+Bytcc=  
49358  sZPVw0KE1FOsLkzG9nSL4InvKt3gvW+GosLlZd2TY1c=  
50335  z0UMCpBGL2FDu0PPbNbD02j6SvTDlXoRpPqzQmpPvF8=  
36982  dqPvRDeeXpl6oAdo1H64dqe0SxDVkQeAw56uV3Z1/oI=  
39963  eJswEvGKsJqPKc+o6eAUB2GY9C2W5Jj+Zl70XUXf+no=  

[5 rows x 157 columns]

In [89]:
X=X_sample[['number_of_days_201702_listened','is_auto_renew','total_cancel','active_days','avg_actual_amount_paid','most_fq_payment_method_id','num_75_mean','number_of_days_listened','num_25_201702_sum','num_50_201702_sum','num_100_201702_sum','num_25_201702_mean','num_50_201702_mean','num_totalsec_lasttwo_mean','total_churn','registered_via_3','registered_via_4','registered_via_7','bd','num_totalsec_sum','num_100_med','num_25_max','num_unq_max','num_totalsec_max','num_totalsec_min','num_985_201702_sum','num_75_201702_mean','num_unq_201702_mean','num_unq_201701_sum','num_25_201612_mean','num_100_201612_mean','num_75_lasttwo_mean','num_50_lastthree_mean','proportion_songs_above_50','registered_via_9']]

In [90]:
y=X_sample['is_churn']

### Tuning optimal number of Neurons

In [93]:
def new_model_nurons(neurons):
    # create model
    model = Sequential()
    model.add(Dense(neurons, input_dim=35, kernel_initializer='normal', activation='relu'))
    #if we want a hidden layer :
    #model.add(Dense(8, activation='relu'))
    model.add(Dense(1, kernel_initializer='normal', activation='sigmoid'))
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

In [94]:
# create model
model = KerasClassifier(build_fn=new_model_nurons, epochs=10, batch_size=16, verbose=1)
# define the grid search parameters
neurons = [1, 5, 10, 15, 20, 25, 30, 50, 90]
param_grid = dict(neurons=neurons)
grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=1)
grid_result = grid.fit(X, y)

Epoch 1/10
4706/4706 [==============================] - 1s 263us/step - loss: 0.6711 - acc: 0.6509
Epoch 2/10
4706/4706 [==============================] - 1s 110us/step - loss: 0.5824 - acc: 0.7524
Epoch 3/10
4706/4706 [==============================] - 1s 114us/step - loss: 0.5210 - acc: 0.7726
Epoch 4/10
4706/4706 [==============================] - 1s 123us/step - loss: 0.4880 - acc: 0.7888
Epoch 5/10
4706/4706 [==============================] - 1s 111us/step - loss: 0.4635 - acc: 0.8032
Epoch 6/10
4706/4706 [==============================] - 1s 119us/step - loss: 0.4271 - acc: 0.8387
Epoch 7/10
4706/4706 [==============================] - 1s 110us/step - loss: 0.3669 - acc: 0.8908
Epoch 8/10
4706/4706 [==============================] - 1s 118us/step - loss: 0.3279 - acc: 0.9052
Epoch 9/10
4706/4706 [==============================] - 1s 129us/step - loss: 0.3042 - acc: 0.9120
Epoch 10/10
4706/4706 [==============================] - 0s 69us/step
Epoch 1/10
4707/4707 [=================

4707/4707 [==============================] - 1s 123us/step - loss: 0.2177 - acc: 0.9216
Epoch 10/10
4707/4707 [==============================] - 0s 72us/step
Epoch 1/10
4706/4706 [==============================] - 2s 346us/step - loss: 0.4528 - acc: 0.8381
Epoch 2/10
4706/4706 [==============================] - 1s 113us/step - loss: 0.2603 - acc: 0.9101
Epoch 3/10
4706/4706 [==============================] - 1s 113us/step - loss: 0.2340 - acc: 0.9197
Epoch 4/10
4706/4706 [==============================] - 1s 125us/step - loss: 0.2259 - acc: 0.9222
Epoch 5/10
4706/4706 [==============================] - 1s 113us/step - loss: 0.2206 - acc: 0.9216
Epoch 6/10
4706/4706 [==============================] - 1s 116us/step - loss: 0.2172 - acc: 0.9235
Epoch 7/10
4706/4706 [==============================] - 1s 113us/step - loss: 0.2142 - acc: 0.9222
Epoch 8/10
4706/4706 [==============================] - 1s 114us/step - loss: 0.2119 - acc: 0.9237
Epoch 9/10
4706/4706 [============================

In [95]:
# summarize results
print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print("%f (%f) with: %r" % (mean, stdev, param))

Best: 0.915722 using {'neurons': 50}
0.913598 (0.008167) with: {'neurons': 1}
0.911756 (0.008153) with: {'neurons': 5}
0.914589 (0.005568) with: {'neurons': 10}
0.913739 (0.007296) with: {'neurons': 15}
0.914306 (0.007792) with: {'neurons': 20}
0.915156 (0.005858) with: {'neurons': 25}
0.915297 (0.004127) with: {'neurons': 30}
0.915722 (0.008514) with: {'neurons': 50}
0.914164 (0.009006) with: {'neurons': 90}


### Tuning batch size and number of training epochs

In [98]:
def new_model():
    # create model
    model = Sequential()
    model.add(Dense(50, input_dim=35, kernel_initializer='normal', activation='relu'))
    #if we want a hidden layer :
    #model.add(Dense(8, activation='relu'))
    model.add(Dense(1, kernel_initializer='normal', activation='sigmoid'))
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

In [99]:
# create model
model = KerasClassifier(build_fn=new_model, verbose=1)
# define the grid search parameters
batch_size = [10,20,30,40,60]
epochs = [10,20,30]
param_grid = dict(batch_size=batch_size, epochs=epochs)
grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=1)
grid_result = grid.fit(X, y)

Epoch 1/10
4706/4706 [==============================] - 2s 497us/step - loss: 0.3630 - acc: 0.8608
Epoch 2/10
4706/4706 [==============================] - 1s 214us/step - loss: 0.2339 - acc: 0.9154
Epoch 3/10
4706/4706 [==============================] - 1s 210us/step - loss: 0.2230 - acc: 0.9184
Epoch 4/10
4706/4706 [==============================] - 1s 214us/step - loss: 0.2186 - acc: 0.9205
Epoch 5/10
4706/4706 [==============================] - 1s 200us/step - loss: 0.2135 - acc: 0.9231
Epoch 6/10
4706/4706 [==============================] - 1s 201us/step - loss: 0.2103 - acc: 0.9237
Epoch 7/10
4706/4706 [==============================] - 1s 202us/step - loss: 0.2069 - acc: 0.9261
Epoch 8/10
4706/4706 [==============================] - 1s 205us/step - loss: 0.2034 - acc: 0.9275
Epoch 9/10
4706/4706 [==============================] - 1s 226us/step - loss: 0.2008 - acc: 0.9258 0s - loss: 0.
Epoch 10/10
4706/4706 [==============================] - 1s 130us/step
Epoch 1/10
4707/4707 [==

4707/4707 [==============================] - 1s 265us/step - loss: 0.2196 - acc: 0.9195
Epoch 6/30
4707/4707 [==============================] - 1s 240us/step - loss: 0.2157 - acc: 0.9188
Epoch 7/30
4707/4707 [==============================] - 1s 272us/step - loss: 0.2126 - acc: 0.9205
Epoch 8/30
4707/4707 [==============================] - 1s 237us/step - loss: 0.2094 - acc: 0.9229
Epoch 9/30
4707/4707 [==============================] - 1s 306us/step - loss: 0.2074 - acc: 0.9252
Epoch 10/30
4707/4707 [==============================] - 1s 266us/step - loss: 0.2031 - acc: 0.9273
Epoch 11/30
4707/4707 [==============================] - 1s 247us/step - loss: 0.1999 - acc: 0.9290
Epoch 12/30
4707/4707 [==============================] - 1s 229us/step - loss: 0.1974 - acc: 0.9265
Epoch 13/30
4707/4707 [==============================] - 1s 278us/step - loss: 0.1950 - acc: 0.9293
Epoch 14/30
4707/4707 [==============================] - 1s 231us/step - loss: 0.1923 - acc: 0.9295
Epoch 15/30
4707

4707/4707 [==============================] - 1s 132us/step - loss: 0.2199 - acc: 0.9201
Epoch 10/30
4707/4707 [==============================] - 1s 109us/step - loss: 0.2167 - acc: 0.9214
Epoch 11/30
4707/4707 [==============================] - 0s 105us/step - loss: 0.2142 - acc: 0.9199
Epoch 12/30
4707/4707 [==============================] - 1s 112us/step - loss: 0.2111 - acc: 0.9239
Epoch 13/30
4707/4707 [==============================] - 1s 115us/step - loss: 0.2089 - acc: 0.9244
Epoch 14/30
4707/4707 [==============================] - 1s 114us/step - loss: 0.2051 - acc: 0.9265
Epoch 15/30
4707/4707 [==============================] - 1s 107us/step - loss: 0.2041 - acc: 0.9246
Epoch 16/30
4707/4707 [==============================] - 1s 110us/step - loss: 0.2023 - acc: 0.9278
Epoch 17/30
4707/4707 [==============================] - 1s 109us/step - loss: 0.1995 - acc: 0.9273
Epoch 18/30
4707/4707 [==============================] - 1s 110us/step - loss: 0.1974 - acc: 0.9273
Epoch 19/30


4706/4706 [==============================] - 0s 67us/step - loss: 0.1984 - acc: 0.9295
Epoch 15/30
4706/4706 [==============================] - 0s 68us/step - loss: 0.1969 - acc: 0.9288
Epoch 16/30
4706/4706 [==============================] - 0s 67us/step - loss: 0.1944 - acc: 0.9295
Epoch 17/30
4706/4706 [==============================] - 0s 68us/step - loss: 0.1929 - acc: 0.9299
Epoch 18/30
4706/4706 [==============================] - 0s 75us/step - loss: 0.1921 - acc: 0.9286
Epoch 19/30
4706/4706 [==============================] - 0s 83us/step - loss: 0.1888 - acc: 0.9309
Epoch 20/30
4706/4706 [==============================] - 0s 91us/step - loss: 0.1874 - acc: 0.9335
Epoch 21/30
4706/4706 [==============================] - 0s 81us/step - loss: 0.1856 - acc: 0.9326
Epoch 22/30
4706/4706 [==============================] - 0s 96us/step - loss: 0.1845 - acc: 0.9337
Epoch 23/30
4706/4706 [==============================] - 0s 71us/step - loss: 0.1816 - acc: 0.9350
Epoch 24/30
4706/4706 

4707/4707 [==============================] - 0s 68us/step - loss: 0.2153 - acc: 0.9208
Epoch 11/20
4707/4707 [==============================] - 0s 68us/step - loss: 0.2145 - acc: 0.9199
Epoch 12/20
4707/4707 [==============================] - 0s 70us/step - loss: 0.2120 - acc: 0.9214
Epoch 13/20
4707/4707 [==============================] - 0s 69us/step - loss: 0.2109 - acc: 0.9216
Epoch 14/20
4707/4707 [==============================] - 0s 69us/step - loss: 0.2093 - acc: 0.9239
Epoch 15/20
4707/4707 [==============================] - 0s 68us/step - loss: 0.2090 - acc: 0.9244
Epoch 16/20
4707/4707 [==============================] - 0s 73us/step - loss: 0.2069 - acc: 0.9263
Epoch 17/20
4707/4707 [==============================] - 0s 70us/step - loss: 0.2057 - acc: 0.9269
Epoch 18/20
4707/4707 [==============================] - 0s 69us/step - loss: 0.2039 - acc: 0.9235
Epoch 19/20
4707/4707 [==============================] - 0s 68us/step - loss: 0.2034 - acc: 0.9248
Epoch 20/20
4707/4707 

4707/4707 [==============================] - 0s 42us/step - loss: 0.2412 - acc: 0.9120
Epoch 7/20
4707/4707 [==============================] - 0s 44us/step - loss: 0.2363 - acc: 0.9133
Epoch 8/20
4707/4707 [==============================] - 0s 59us/step - loss: 0.2337 - acc: 0.9127
Epoch 9/20
4707/4707 [==============================] - 0s 40us/step - loss: 0.2304 - acc: 0.9157
Epoch 10/20
4707/4707 [==============================] - 0s 42us/step - loss: 0.2277 - acc: 0.9188
Epoch 11/20
4707/4707 [==============================] - 0s 41us/step - loss: 0.2259 - acc: 0.9205
Epoch 12/20
4707/4707 [==============================] - 0s 41us/step - loss: 0.2229 - acc: 0.9201
Epoch 13/20
4707/4707 [==============================] - 0s 42us/step - loss: 0.2203 - acc: 0.9218
Epoch 14/20
4707/4707 [==============================] - 0s 41us/step - loss: 0.2187 - acc: 0.9218
Epoch 15/20
4707/4707 [==============================] - 0s 41us/step - loss: 0.2175 - acc: 0.9216
Epoch 16/20
4707/4707 [==

In [100]:
# summarize results
print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print("%f (%f) with: %r" % (mean, stdev, param))

Best: 0.916856 using {'batch_size': 40, 'epochs': 30}
0.914164 (0.006410) with: {'batch_size': 10, 'epochs': 10}
0.916431 (0.007130) with: {'batch_size': 10, 'epochs': 20}
0.915864 (0.006064) with: {'batch_size': 10, 'epochs': 30}
0.914589 (0.004678) with: {'batch_size': 20, 'epochs': 10}
0.915156 (0.006960) with: {'batch_size': 20, 'epochs': 20}
0.914448 (0.007476) with: {'batch_size': 20, 'epochs': 30}
0.914873 (0.007077) with: {'batch_size': 30, 'epochs': 10}
0.914873 (0.009388) with: {'batch_size': 30, 'epochs': 20}
0.916006 (0.006435) with: {'batch_size': 30, 'epochs': 30}
0.914164 (0.007156) with: {'batch_size': 40, 'epochs': 10}
0.915297 (0.007879) with: {'batch_size': 40, 'epochs': 20}
0.916856 (0.007133) with: {'batch_size': 40, 'epochs': 30}
0.912465 (0.007936) with: {'batch_size': 60, 'epochs': 10}
0.911898 (0.007214) with: {'batch_size': 60, 'epochs': 20}
0.915297 (0.005711) with: {'batch_size': 60, 'epochs': 30}


### Tuning Training Optimization Algorithm

In [101]:
def new_model_opt(optimizer):
    # create model
    model = Sequential()
    model.add(Dense(50, input_dim=35, kernel_initializer='normal', activation='relu'))
    #if we want a hidden layer :
    #model.add(Dense(8, activation='relu'))
    model.add(Dense(1, kernel_initializer='normal', activation='sigmoid'))
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])
    return model

In [102]:
# create model
model = KerasClassifier(build_fn=new_model_opt, epochs=30, batch_size=40, verbose=1)
# define the grid search parameters
optimizer = ['SGD', 'RMSprop', 'Adagrad', 'Adadelta', 'Adam', 'Adamax', 'Nadam']
param_grid = dict(optimizer=optimizer)
grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=1)
grid_result = grid.fit(X, y)


Epoch 1/30
4706/4706 [==============================] - 3s 565us/step - loss: 0.6820 - acc: 0.6477
Epoch 2/30
4706/4706 [==============================] - 0s 64us/step - loss: 0.6273 - acc: 0.8330
Epoch 3/30
4706/4706 [==============================] - 0s 64us/step - loss: 0.5522 - acc: 0.8572
Epoch 4/30
4706/4706 [==============================] - 0s 63us/step - loss: 0.4758 - acc: 0.8731
Epoch 5/30
4706/4706 [==============================] - 0s 69us/step - loss: 0.4144 - acc: 0.8878
Epoch 6/30
4706/4706 [==============================] - 0s 62us/step - loss: 0.3680 - acc: 0.8940
Epoch 7/30
4706/4706 [==============================] - 0s 63us/step - loss: 0.3345 - acc: 0.8995
Epoch 8/30
4706/4706 [==============================] - 0s 71us/step - loss: 0.3105 - acc: 0.9014
Epoch 9/30
4706/4706 [==============================] - 0s 65us/step - loss: 0.2933 - acc: 0.9031
Epoch 10/30
4706/4706 [==============================] - 0s 63us/step - loss: 0.2806 - acc: 0.9046
Epoch 11/30
4706/4

4707/4707 [==============================] - 0s 64us/step - loss: 0.2188 - acc: 0.9212
Epoch 10/30
4707/4707 [==============================] - 0s 82us/step - loss: 0.2168 - acc: 0.9222
Epoch 11/30
4707/4707 [==============================] - 0s 73us/step - loss: 0.2149 - acc: 0.9231
Epoch 12/30
4707/4707 [==============================] - 0s 61us/step - loss: 0.2136 - acc: 0.9239
Epoch 13/30
4707/4707 [==============================] - 0s 62us/step - loss: 0.2116 - acc: 0.9237
Epoch 14/30
4707/4707 [==============================] - 0s 60us/step - loss: 0.2099 - acc: 0.9252
Epoch 15/30
4707/4707 [==============================] - 0s 74us/step - loss: 0.2082 - acc: 0.9259
Epoch 16/30
4707/4707 [==============================] - 0s 71us/step - loss: 0.2065 - acc: 0.9276
Epoch 17/30
4707/4707 [==============================] - 0s 69us/step - loss: 0.2050 - acc: 0.9278
Epoch 18/30
4707/4707 [==============================] - 0s 65us/step - loss: 0.2033 - acc: 0.9273
Epoch 19/30
4707/4707 

4707/4707 [==============================] - 0s 59us/step - loss: 0.2292 - acc: 0.9163
Epoch 19/30
4707/4707 [==============================] - 0s 68us/step - loss: 0.2282 - acc: 0.9161
Epoch 20/30
4707/4707 [==============================] - 0s 72us/step - loss: 0.2274 - acc: 0.9167
Epoch 21/30
4707/4707 [==============================] - 0s 63us/step - loss: 0.2268 - acc: 0.9169
Epoch 22/30
4707/4707 [==============================] - 0s 60us/step - loss: 0.2258 - acc: 0.9171
Epoch 23/30
4707/4707 [==============================] - 0s 60us/step - loss: 0.2250 - acc: 0.9171
Epoch 24/30
4707/4707 [==============================] - 0s 59us/step - loss: 0.2245 - acc: 0.9174
Epoch 25/30
4707/4707 [==============================] - 0s 63us/step - loss: 0.2237 - acc: 0.9171
Epoch 26/30
4707/4707 [==============================] - 0s 70us/step - loss: 0.2230 - acc: 0.9184
Epoch 27/30
4707/4707 [==============================] - 0s 64us/step - loss: 0.2222 - acc: 0.9182
Epoch 28/30
4707/4707 

4706/4706 [==============================] - 0s 61us/step - loss: 0.1958 - acc: 0.9278
Epoch 28/30
4706/4706 [==============================] - 0s 80us/step - loss: 0.1949 - acc: 0.9292
Epoch 29/30
4706/4706 [==============================] - 0s 76us/step - loss: 0.1937 - acc: 0.9286
Epoch 30/30
4706/4706 [==============================] - 0s 55us/step
Epoch 1/30
4707/4707 [==============================] - 3s 730us/step - loss: 0.5204 - acc: 0.7931
Epoch 2/30
4707/4707 [==============================] - 0s 82us/step - loss: 0.3299 - acc: 0.8885
Epoch 3/30
4707/4707 [==============================] - 0s 71us/step - loss: 0.2749 - acc: 0.9018
Epoch 4/30
4707/4707 [==============================] - 0s 62us/step - loss: 0.2574 - acc: 0.9076
Epoch 5/30
4707/4707 [==============================] - 0s 75us/step - loss: 0.2489 - acc: 0.9101
Epoch 6/30
4707/4707 [==============================] - 0s 62us/step - loss: 0.2436 - acc: 0.9112
Epoch 7/30
4707/4707 [==============================] - 

7060/7060 [==============================] - 0s 66us/step - loss: 0.2288 - acc: 0.9156
Epoch 6/30
7060/7060 [==============================] - 0s 65us/step - loss: 0.2256 - acc: 0.9180
Epoch 7/30
7060/7060 [==============================] - 0s 69us/step - loss: 0.2225 - acc: 0.9186
Epoch 8/30
7060/7060 [==============================] - 1s 79us/step - loss: 0.2213 - acc: 0.9211
Epoch 9/30
7060/7060 [==============================] - 1s 76us/step - loss: 0.2186 - acc: 0.9203
Epoch 10/30
7060/7060 [==============================] - 0s 67us/step - loss: 0.2156 - acc: 0.9224
Epoch 11/30
7060/7060 [==============================] - 1s 79us/step - loss: 0.2139 - acc: 0.9237
Epoch 12/30
7060/7060 [==============================] - 1s 78us/step - loss: 0.2118 - acc: 0.9248
Epoch 13/30
7060/7060 [==============================] - 1s 72us/step - loss: 0.2101 - acc: 0.9251
Epoch 14/30
7060/7060 [==============================] - 1s 78us/step - loss: 0.2075 - acc: 0.9256
Epoch 15/30
7060/7060 [===

In [103]:
# summarize results
print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print("%f (%f) with: %r" % (mean, stdev, param))

Best: 0.917564 using {'optimizer': 'Adam'}
0.911473 (0.008547) with: {'optimizer': 'SGD'}
0.917564 (0.006231) with: {'optimizer': 'RMSprop'}
0.914164 (0.006605) with: {'optimizer': 'Adagrad'}
0.913598 (0.008215) with: {'optimizer': 'Adadelta'}
0.917564 (0.006243) with: {'optimizer': 'Adam'}
0.914589 (0.007988) with: {'optimizer': 'Adamax'}
0.913739 (0.005718) with: {'optimizer': 'Nadam'}


### Tuning Neuron Activation Function

In [104]:
def new_model_act(activation):
    # create model
    model = Sequential()
    model.add(Dense(50, input_dim=35, kernel_initializer='normal', activation=activation))
    #if we want a hidden layer :
    #model.add(Dense(8, activation='relu'))
    model.add(Dense(1, kernel_initializer='normal', activation='sigmoid'))
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='Adam', metrics=['accuracy'])
    return model

In [105]:
# create model
model = KerasClassifier(build_fn=new_model_act, epochs=30, batch_size=40, verbose=1)
# define the grid search parameters
activation = ['softmax', 'softplus', 'softsign', 'relu', 'tanh', 'sigmoid', 'hard_sigmoid', 'linear']
param_grid = dict(activation=activation)
grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=1)
grid_result = grid.fit(X, y)


Epoch 1/30
4706/4706 [==============================] - 3s 709us/step - loss: 0.6829 - acc: 0.7384
Epoch 2/30
4706/4706 [==============================] - 0s 82us/step - loss: 0.6414 - acc: 0.8417
Epoch 3/30
4706/4706 [==============================] - 0s 73us/step - loss: 0.5806 - acc: 0.8825
Epoch 4/30
4706/4706 [==============================] - 0s 80us/step - loss: 0.5181 - acc: 0.8991
Epoch 5/30
4706/4706 [==============================] - 0s 72us/step - loss: 0.4668 - acc: 0.9067
Epoch 6/30
4706/4706 [==============================] - 0s 74us/step - loss: 0.4271 - acc: 0.9101
Epoch 7/30
4706/4706 [==============================] - 0s 79us/step - loss: 0.3961 - acc: 0.9137: 0s - loss: 0.4115 - acc: 
Epoch 8/30
4706/4706 [==============================] - 0s 77us/step - loss: 0.3707 - acc: 0.9176
Epoch 9/30
4706/4706 [==============================] - 0s 80us/step - loss: 0.3499 - acc: 0.9193
Epoch 10/30
4706/4706 [==============================] - 0s 77us/step - loss: 0.3322 - acc

4707/4707 [==============================] - 0s 75us/step - loss: 0.2311 - acc: 0.9159
Epoch 11/30
4707/4707 [==============================] - 0s 73us/step - loss: 0.2298 - acc: 0.9131
Epoch 12/30
4707/4707 [==============================] - 0s 68us/step - loss: 0.2286 - acc: 0.9161
Epoch 13/30
4707/4707 [==============================] - 0s 67us/step - loss: 0.2278 - acc: 0.9154
Epoch 14/30
4707/4707 [==============================] - 0s 65us/step - loss: 0.2268 - acc: 0.9157
Epoch 15/30
4707/4707 [==============================] - 0s 66us/step - loss: 0.2258 - acc: 0.9154
Epoch 16/30
4707/4707 [==============================] - 0s 68us/step - loss: 0.2245 - acc: 0.9146
Epoch 17/30
4707/4707 [==============================] - 0s 75us/step - loss: 0.2236 - acc: 0.9174
Epoch 18/30
4707/4707 [==============================] - 0s 72us/step - loss: 0.2234 - acc: 0.9167
Epoch 19/30
4707/4707 [==============================] - 0s 65us/step - loss: 0.2229 - acc: 0.9161
Epoch 20/30
4707/4707 

4707/4707 [==============================] - 0s 65us/step - loss: 0.2051 - acc: 0.9254
Epoch 20/30
4707/4707 [==============================] - 0s 63us/step - loss: 0.2035 - acc: 0.9252
Epoch 21/30
4707/4707 [==============================] - 0s 65us/step - loss: 0.2013 - acc: 0.9250
Epoch 22/30
4707/4707 [==============================] - 0s 65us/step - loss: 0.2001 - acc: 0.9269
Epoch 23/30
4707/4707 [==============================] - 0s 65us/step - loss: 0.1991 - acc: 0.9282
Epoch 24/30
4707/4707 [==============================] - 0s 64us/step - loss: 0.1976 - acc: 0.9263
Epoch 25/30
4707/4707 [==============================] - 0s 64us/step - loss: 0.1955 - acc: 0.9280
Epoch 26/30
4707/4707 [==============================] - 0s 69us/step - loss: 0.1943 - acc: 0.9280
Epoch 27/30
4707/4707 [==============================] - 0s 72us/step - loss: 0.1930 - acc: 0.9295
Epoch 28/30
4707/4707 [==============================] - 0s 64us/step - loss: 0.1910 - acc: 0.9301
Epoch 29/30
4707/4707 

4706/4706 [==============================] - 0s 74us/step - loss: 0.2267 - acc: 0.9197
Epoch 29/30
4706/4706 [==============================] - 0s 71us/step - loss: 0.2262 - acc: 0.9224
Epoch 30/30
4706/4706 [==============================] - 0s 41us/step
Epoch 1/30
4707/4707 [==============================] - 4s 757us/step - loss: 0.6197 - acc: 0.7340
Epoch 2/30
4707/4707 [==============================] - 0s 63us/step - loss: 0.4385 - acc: 0.8443
Epoch 3/30
4707/4707 [==============================] - 0s 62us/step - loss: 0.3292 - acc: 0.8893
Epoch 4/30
4707/4707 [==============================] - 0s 62us/step - loss: 0.2856 - acc: 0.9021
Epoch 5/30
4707/4707 [==============================] - 0s 62us/step - loss: 0.2709 - acc: 0.9031
Epoch 6/30
4707/4707 [==============================] - 0s 62us/step - loss: 0.2650 - acc: 0.9042
Epoch 7/30
4707/4707 [==============================] - 0s 63us/step - loss: 0.2620 - acc: 0.9055
Epoch 8/30
4707/4707 [==============================] - 0

4706/4706 [==============================] - 0s 72us/step - loss: 0.2415 - acc: 0.9167: 0s - loss: 0.2297 - acc: 0
Epoch 7/30
4706/4706 [==============================] - 0s 74us/step - loss: 0.2405 - acc: 0.9186
Epoch 8/30
4706/4706 [==============================] - 0s 71us/step - loss: 0.2411 - acc: 0.9173
Epoch 9/30
4706/4706 [==============================] - 0s 72us/step - loss: 0.2422 - acc: 0.9180
Epoch 10/30
4706/4706 [==============================] - 0s 72us/step - loss: 0.2406 - acc: 0.9180
Epoch 11/30
4706/4706 [==============================] - 0s 70us/step - loss: 0.2403 - acc: 0.9188
Epoch 12/30
4706/4706 [==============================] - 0s 70us/step - loss: 0.2407 - acc: 0.9169
Epoch 13/30
4706/4706 [==============================] - 0s 72us/step - loss: 0.2400 - acc: 0.9186
Epoch 14/30
4706/4706 [==============================] - 0s 70us/step - loss: 0.2407 - acc: 0.9165
Epoch 15/30
4706/4706 [==============================] - 0s 70us/step - loss: 0.2404 - acc: 0.91

In [106]:
# summarize results
print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print("%f (%f) with: %r" % (mean, stdev, param))

Best: 0.918130 using {'activation': 'relu'}
0.913173 (0.010280) with: {'activation': 'softmax'}
0.913739 (0.009300) with: {'activation': 'softplus'}
0.914023 (0.010233) with: {'activation': 'softsign'}
0.918130 (0.005206) with: {'activation': 'relu'}
0.916856 (0.007404) with: {'activation': 'tanh'}
0.914731 (0.007351) with: {'activation': 'sigmoid'}
0.912890 (0.008579) with: {'activation': 'hard_sigmoid'}
0.911190 (0.006001) with: {'activation': 'linear'}


### Testing with Hidden Layer

In [109]:
def new_model_hl():
    # create model
    model = Sequential()
    model.add(Dense(50, input_dim=35, kernel_initializer='normal', activation='relu'))
    #if we want a hidden layer :
    model.add(Dense(8, activation='relu'))
    model.add(Dense(1, kernel_initializer='normal', activation='sigmoid'))
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='Adam', metrics=['accuracy'])
    return model

In [111]:
model=new_model_hl()
model.fit(X_train_no_corr, y_train,epochs=30, batch_size=40, verbose=1)

Epoch 1/30
70598/70598 [==============================] - 9s 129us/step - loss: 0.2550 - acc: 0.9051
Epoch 2/30
70598/70598 [==============================] - 6s 80us/step - loss: 0.2095 - acc: 0.9240
Epoch 3/30
70598/70598 [==============================] - 6s 81us/step - loss: 0.2041 - acc: 0.9255
Epoch 4/30
70598/70598 [==============================] - 6s 85us/step - loss: 0.2011 - acc: 0.9261
Epoch 5/30
70598/70598 [==============================] - 6s 85us/step - loss: 0.1991 - acc: 0.9274
Epoch 6/30
70598/70598 [==============================] - 6s 80us/step - loss: 0.1975 - acc: 0.9272
Epoch 7/30
70598/70598 [==============================] - 6s 80us/step - loss: 0.1959 - acc: 0.9278
Epoch 8/30
70598/70598 [==============================] - 6s 82us/step - loss: 0.1946 - acc: 0.9279
Epoch 9/30
70598/70598 [==============================] - 6s 85us/step - loss: 0.1937 - acc: 0.9283
Epoch 10/30
70598/70598 [==============================] - 6s 83us/step - loss: 0.1924 - acc: 0.928

In [112]:
analytics(y_test,model,X_test_no_corr)

Confusion Matrix:
      Prediction
         0     1
    0220709 24841
    1 1068 16225

Accuracy Score:
0.901427848564
Classification Report:
             precision    recall  f1-score   support

          0       1.00      0.90      0.94    245550
          1       0.40      0.94      0.56     17293

avg / total       0.96      0.90      0.92    262843

Log Loss:
0.214895750254
AUC Score:
0.970539598374


## Final Model according to parameter tuning
Chosen configuration:
- input dimensions: 35 after removing correlated features
- activation function: Relu
- optimizer: Adam
- hidden layer: Yes

In [116]:
def new_model_final():
    # create model
    model = Sequential()
    model.add(Dense(50, input_dim=35, kernel_initializer='normal', activation='relu'))
    #if we want a hidden layer :
    model.add(Dense(8, activation='relu'))
    model.add(Dense(1, kernel_initializer='normal', activation='sigmoid'))
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='Adam', metrics=['accuracy'])
    return model

In [117]:
model_final=new_model_final()
model_final.fit(X_train_no_corr, y_train,epochs=30, batch_size=40, verbose=1)

Epoch 1/30
70598/70598 [==============================] - 10s 140us/step - loss: 0.2503 - acc: 0.9101
Epoch 2/30
70598/70598 [==============================] - 7s 96us/step - loss: 0.2082 - acc: 0.9246
Epoch 3/30
70598/70598 [==============================] - 6s 88us/step - loss: 0.2033 - acc: 0.9264
Epoch 4/30
70598/70598 [==============================] - 6s 88us/step - loss: 0.2004 - acc: 0.9273
Epoch 5/30
70598/70598 [==============================] - 6s 88us/step - loss: 0.1980 - acc: 0.9276
Epoch 6/30
70598/70598 [==============================] - 6s 87us/step - loss: 0.1961 - acc: 0.9282
Epoch 7/30
70598/70598 [==============================] - 6s 80us/step - loss: 0.1949 - acc: 0.9281
Epoch 8/30
70598/70598 [==============================] - 6s 80us/step - loss: 0.1932 - acc: 0.9286
Epoch 9/30
70598/70598 [==============================] - 6s 90us/step - loss: 0.1919 - acc: 0.9287
Epoch 10/30
70598/70598 [==============================] - 5s 74us/step - loss: 0.1909 - acc: 0.92

In [118]:
analytics(y_test,model_final,X_test_no_corr)

Confusion Matrix:
      Prediction
         0     1
    0222852 22698
    1 1242 16051

Accuracy Score:
0.90891901249
Classification Report:
             precision    recall  f1-score   support

          0       0.99      0.91      0.95    245550
          1       0.41      0.93      0.57     17293

avg / total       0.96      0.91      0.92    262843

Log Loss:


/Users/sdoneva/anaconda/lib/python3.6/site-packages/sklearn/metrics/classification.py:1694: RuntimeWarning: divide by zero encountered in log
  loss = -(transformed_labels * np.log(y_pred)).sum(axis=1)
/Users/sdoneva/anaconda/lib/python3.6/site-packages/sklearn/metrics/classification.py:1694: RuntimeWarning: invalid value encountered in multiply
  loss = -(transformed_labels * np.log(y_pred)).sum(axis=1)


nan
AUC Score:
0.970319632205


### Statistics

In [12]:
import statsmodels.api as sm
from sklearn import linear_model
from sklearn import preprocessing

def stat_logit(data_target,data_indepen):
    logit_model=sm.Logit(data_target,data_indepen)
    result=logit_model.fit()
    return print(result.summary())

def train_logistic_regression(data_target,data_indepen):
    # Initialize logistic regression model
    log_model = linear_model.LogisticRegression()
    # Train the model
    log_model.fit(X =data_indepen,y = data_target)
    return log_model

from sklearn.metrics import confusion_matrix
from sklearn.utils.multiclass import unique_labels
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

def confusion_matrix_report(y_true, y_pred):    
    cm, labels = confusion_matrix(y_true, y_pred), unique_labels(y_true, y_pred)
    column_width = max([len(str(x)) for x in labels] + [5])  # 5 is value length
    report = " " * column_width + " " + "{:_^{}}".format("Prediction", column_width * len(labels))+ "\n"
    report += " " * column_width + " ".join(["{:>{}}".format(label, column_width) for label in labels]) + "\n"
    for i, label1 in enumerate(labels):
        report += "{:>{}}".format(label1, column_width) + " ".join(["{:{}d}".format(cm[i, j], column_width) for j in range(len(labels))]) + "\n"
    return report

def loggloss(target_test, model, data_test):
    probabilities=model.predict_proba(data_test)
    value=log_loss(target_test, probabilities)
    return value

def AUC(target_test, model, data_test):
    values=model.predict_proba(data_test) #[:,1]
    auc_score=roc_auc_score(target_test, values)
    return auc_score
    

def analytics(target_test, model, data_test):#target of the test data #predictions as 0,1 #model (classifier) #data_test
    y_pred=model.predict_classes(data_test) #changed for ANN
    print("Confusion Matrix:")
    print(confusion_matrix_report(target_test,y_pred))
    print("Accuracy Score:")
    print(accuracy_score(target_test,y_pred))
    print("Classification Report:")
    print(classification_report(target_test,y_pred))
    print("Log Loss:")
    print(loggloss(target_test, model, data_test))
    print("AUC Score:")
    print(AUC(target_test, model, data_test))

/Users/sdoneva/anaconda/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools
